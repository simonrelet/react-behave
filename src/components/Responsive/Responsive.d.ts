import * as React from 'react'

export interface ResponsiveProps {
  children: (
    screenSize: string,
    width: number,
  ) => React.ReactNode | React.ReactNode
  maximum?: string
  minimum?: string
  resizeInterval?: number
  screenSizes: { [key: string]: number }
}

declare const Responsive: React.ComponentType<ResponsiveProps>

export default Responsive
