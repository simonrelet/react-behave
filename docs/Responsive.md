<!--
  THIS FILE WAS GENERATED!
  Don't make any changes in it, update src/components/Responsive/Responsive.js instead.
  If you still need to make changes in this file, remove this header so it won't be overridden.
-->

# Responsive

[props-minimum]: #minimum-string-optional
[props-maximum]: #maximum-string-optional
[props-screensizes]: #screensizes-object-optional

Render content depending on the screen size.

## Usage

```jsx
import React, { Component } from 'react'
import { Responsive } from 'react-behave'

const screenSizes = {
  xs: 0,
  sm: 600,
  md: 960,
  lg: 1280,
  xl: 1920,
}

class App extends Component {
  render() {
    return (
      <Responsive minimum="md" screenSizes={screenSizes}>
        <p>Only visible on medium and bigger screens</p>
      </Responsive>
    )
  }
}
```

## Props

### `children`: `Function`|`Node`

_Parameters_: `screenSize: String`, `width: Number`

Invoked to render the children.

The children will only be rendered if the current width is satisfied by [`props.maximum`][props-maximum] and [`props.minimum`][props-minimum].

Example:

```jsx
import React from 'react'
import { Responsive } from 'react-behave'

const screenSizes = {
  // [...]
}

function LargeScreen() {
  return (
    <Responsive minimum="lg" screenSizes={screenSizes}>
      {screenSize => <p>The screen size is '{screenSize}'</p>}
    </Responsive>
  )
}
```

For convenience, `props.children` can also be a React element.
See [Usage](#usage).

### `maximum`: `String` (optional)

Maximum screen width.
Must be one of the keys of [`props.screenSizes`][props-screensizes].

### `minimum`: `String` (optional)

Minimum screen width.
Must be one of the keys of [`props.screenSizes`][props-screensizes].

### `resizeInterval`: `Number` (optional)

_Default value_: `166`

The minimum interval between two resizes.

### `screenSizes`: `Object` (optional)

_Default value_:

```jsx
{
  xs: 0,
  sm: 600,
  md: 960,
  lg: 1280,
  xl: 1920,
}
```

The screen sizes to use.
Each key is the name of a screen size that will be used by [`props.maximum`][props-maximum] and [`props.minimum`][props-minimum], and the value is the starting width in pixels of this range.

For example:

```js
const screenSizes = {
  sm: 0,
  md: 960,
  lg: 1280,
}
```

Define the following screen sizes:

- `'sm'`: [0, 960[
- `'md'`: [960, 1280[
- `'lg'`: [1280, âˆž[
