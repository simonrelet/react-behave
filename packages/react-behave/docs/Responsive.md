# Responsive

> Render content depending on the screen size.

## Usage

```jsx
import React, { Component } from 'react';
import { Responsive } from 'react-behave';

const screenSizes = {
  xs: 0,
  sm: 600,
  md: 960,
  lg: 1280,
  xl: 1920,
};

class App extends Component {
  render() {
    return (
      <Responsive minimum="md" screenSizes={screenSizes}>
        <p>Only visible on medium and bigger screens</p>
      </Responsive>
    );
  }
}
```

## Screen sizes

The screen sizes are defined in an object.
Each key is the name of a screen size that will be used by `maximum` and `minimum` props and the value is the starting width in pixels of this range.

For example:

```js
const screenSizes = {
  sm: 0,
  md: 960,
  lg: 1280,
};
```

Define the following screen sizes:

- `sm`: [0, 960[
- `md`: [960, 1280[
- `lg`: [1280, ∞[

## Render methods

### `<Responsive children />`

The children will only be rendered if the current width is satisfied by `maximum` and `minimum`.

See the [Usage](#usage) for an example.

### `<Responsive render />`

The render function will only be called if the current width is satisfied by `maximum` and `minimum`.
This function takes the width as first parameter.

Warning: `<Responsive children />` takes precedence over `<Responsive render />` so don’t use both.

Example:

```jsx
const screenSizes = {
  // [...]
};

function LargeScreen() {
  return (
    <Responsive
      minimum="lg"
      screenSizes={screenSizes}
      render={width => <p>The width is '{width}'</p>}
    />
  );
}
```

## Props

_Mandatory props are marked with a `*`._

| Name                                        |   Type   | Default value | Description                                                          |
| ------------------------------------------- | :------: | :-----------: | -------------------------------------------------------------------- |
| `children`                                  |  `node`  |               | Children to render. See the [render methods](#responsive-children-). |
| `maximum`                                   | `string` |               | Maximum screen width. Must be one of the `screenSizes`.              |
| `minimum`                                   | `string` |               | Minimum screen width. Must be one of the `screenSizes`.              |
| `render`                                    |  `func`  |               | Render function. See the [render methods](#responsive-render-).      |
| <strong><code>screenSizes</code>\*</strong> | `object` |               | Screen sizes to use. See [Screen sizes](#screen-sizes).              |
