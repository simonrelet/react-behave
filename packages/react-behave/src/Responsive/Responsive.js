import { renderMethod } from '@simonrelet/react-utils';
import debounce from 'lodash.debounce';
import isEqual from 'lodash.isequal';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import EventListener from 'react-event-listener';

// Corresponds to 10 frames at 60 Hz
const RESIZE_INTERVAL = 166;

function smallerOrEqualTo(screenSizes, refWidth, width) {
  return screenSizes[width] <= screenSizes[refWidth];
}

function biggerOrEqualTo(screenSizes, refWidth, width) {
  return screenSizes[width] >= screenSizes[refWidth];
}

function getScreenWidth(screenSizes, windowWidth) {
  return Object.keys(screenSizes)
    .reverse()
    .find(screenSizeName => windowWidth >= screenSizes[screenSizeName]);
}

/**
 * > Render content depending on the screen size.
 *
 * ## Usage
 *
 * ```jsx
 * import React, { Component } from 'react';
 * import { Responsive } from 'react-behave';
 *
 * const screenSizes = {
 *   xs: 0,
 *   sm: 600,
 *   md: 960,
 *   lg: 1280,
 *   xl: 1920,
 * };
 *
 * class App extends Component {
 *   render() {
 *     <Responsive up="md" screenSizes={screenSizes}>
 *       <p>Only visible on medium and bigger screens</p>
 *     </Responsive>
 *   }
 * }
 * ```
 *
 * ## Screen sizes
 *
 * The screen sizes are defined in an object.
 * Each key is the name of a screen size that will be used by `down` and `up` props and the value is the starting width in pixels of this range.
 *
 * For example:
 *
 * ```js
 * const screenSizes = {
 *   sm: 0,
 *   md: 960,
 *   lg: 1280,
 * };
 * ```
 *
 * Define the following screen sizes:
 *
 * - `sm`: [0, 960[
 * - `md`: [960, 1280[
 * - `lg`: [1280, ∞[
 */
class Responsive extends Component {
  static propTypes = {
    /**
     * Children to render if the current width is satisfied by `down` and `up`.
     */
    children: PropTypes.node,

    /**
     * Maximum screen width.
     * Must be one of the `screenSizes`.
     */
    down: PropTypes.string,

    /**
     * Render function.
     * Warning: `children` takes precedence over `render` so don’t use both.
     */
    render: PropTypes.func,

    /**
     * Screen sizes to use.
     * See [Screen sizes](#screen-sizes).
     */
    screenSizes: PropTypes.object.isRequired,

    /**
     * Minimum screen width.
     * Must be one of the `screenSizes`.
     */
    up: PropTypes.string,
  };

  state = {
    screenSizes: null,
    width: null,
  };

  static getDerivedStateFromProps(props, state) {
    if (!isEqual(props.screenSizes, state.screenSizes)) {
      return {
        screenSizes: props.screenSizes,
        width: getScreenWidth(props.screenSizes, window.innerWidth),
      };
    }
    return null;
  }

  componentWillUnmount() {
    this.handleResize.cancel();
  }

  updateWidth = () => {
    const width = getScreenWidth(this.props.screenSizes, window.innerWidth);

    if (width !== this.state.width) {
      this.setState({ width });
    }
  };

  handleResize = debounce(this.updateWidth, RESIZE_INTERVAL);

  render() {
    const { width } = this.state;
    const { children, down, render, screenSizes, up } = this.props;

    if (!width) {
      return null;
    }

    let visible = true;

    if (up) {
      visible = biggerOrEqualTo(screenSizes, up, width);
    }

    if (visible && down) {
      visible = smallerOrEqualTo(screenSizes, down, width);
    }

    return (
      <EventListener target="window" onResize={this.handleResize}>
        {visible && renderMethod({ children, render }, width)}
      </EventListener>
    );
  }
}

export default Responsive;
