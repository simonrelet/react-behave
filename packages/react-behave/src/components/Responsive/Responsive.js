import debounce from 'lodash.debounce';
import isEqual from 'lodash.isequal';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import EventListener from 'react-event-listener';
import renderMethod from '../../core/renderMethod';

// Corresponds to 10 frames at 60 Hz
// Delay to next tick for tests.
const RESIZE_INTERVAL = process.env.NODE_ENV === 'test' ? 0 : 166;

function smallerOrEqualTo(screenSizes, refWidth, width) {
  return screenSizes[width] <= screenSizes[refWidth];
}

function biggerOrEqualTo(screenSizes, refWidth, width) {
  return screenSizes[width] >= screenSizes[refWidth];
}

function getScreenWidth(screenSizes, windowWidth) {
  return Object.keys(screenSizes)
    .reverse()
    .find(screenSizeName => windowWidth >= screenSizes[screenSizeName]);
}

/**
 * > Render content depending on the screen size.
 *
 * ## Usage
 *
 * ```jsx
 * import React, { Component } from 'react';
 * import { Responsive } from 'react-behave';
 *
 * const screenSizes = {
 *   xs: 0,
 *   sm: 600,
 *   md: 960,
 *   lg: 1280,
 *   xl: 1920,
 * };
 *
 * class App extends Component {
 *   render() {
 *     return (
 *       <Responsive minimum="md" screenSizes={screenSizes}>
 *         <p>Only visible on medium and bigger screens</p>
 *       </Responsive>
 *     );
 *   }
 * }
 * ```
 *
 * ## Screen sizes
 *
 * The screen sizes are defined in an object.
 * Each key is the name of a screen size that will be used by `maximum` and `minimum` props and the value is the starting width in pixels of this range.
 *
 * For example:
 *
 * ```js
 * const screenSizes = {
 *   sm: 0,
 *   md: 960,
 *   lg: 1280,
 * };
 * ```
 *
 * Define the following screen sizes:
 *
 * - `sm`: [0, 960[
 * - `md`: [960, 1280[
 * - `lg`: [1280, ∞[
 *
 * ## Render methods
 *
 * ### `<Responsive children />`
 *
 * The children will only be rendered if the current width is satisfied by `maximum` and `minimum`.
 *
 * See the [Usage](#usage) for an example.
 *
 * ### `<Responsive render />`
 *
 * The render function will only be called if the current width is satisfied by `maximum` and `minimum`.
 * This function takes the width as first parameter.
 *
 * Warning: `<Responsive children />` takes precedence over `<Responsive render />` so don’t use both.
 *
 * Example:
 *
 * ```jsx
 * const screenSizes = {
 *   // [...]
 * };
 *
 * function LargeScreen() {
 *   return (
 *     <Responsive
 *       minimum="lg"
 *       screenSizes={screenSizes}
 *       render={width => <p>The width is '{width}'</p>}
 *     />
 *   );
 * }
 * ```
 */
class Responsive extends Component {
  static propTypes = {
    /**
     * Children to render.
     * See the [render methods](#responsive-children-).
     */
    children: PropTypes.node,

    /**
     * Maximum screen width.
     * Must be one of the `screenSizes`.
     */
    maximum: PropTypes.string,

    /**
     * Minimum screen width.
     * Must be one of the `screenSizes`.
     */
    minimum: PropTypes.string,

    /**
     * Render function.
     * See the [render methods](#responsive-render-).
     */
    render: PropTypes.func,

    /**
     * Screen sizes to use.
     * See [Screen sizes](#screen-sizes).
     */
    screenSizes: PropTypes.object.isRequired,
  };

  state = {
    screenSizes: null,
    width: null,
  };

  static getDerivedStateFromProps(props, state) {
    if (!isEqual(props.screenSizes, state.screenSizes)) {
      return {
        screenSizes: props.screenSizes,
        width: getScreenWidth(props.screenSizes, window.innerWidth),
      };
    }
    return null;
  }

  componentWillUnmount() {
    this.handleResize.cancel();
  }

  updateWidth = () => {
    const width = getScreenWidth(this.props.screenSizes, window.innerWidth);

    if (width !== this.state.width) {
      this.setState({ width });
    }
  };

  handleResize = debounce(this.updateWidth, RESIZE_INTERVAL);

  render() {
    const { width } = this.state;
    const { children, maximum, minimum, render, screenSizes } = this.props;

    let visible = true;

    if (minimum) {
      visible = biggerOrEqualTo(screenSizes, minimum, width);
    }

    if (visible && maximum) {
      visible = smallerOrEqualTo(screenSizes, maximum, width);
    }

    return (
      <EventListener target="window" onResize={this.handleResize}>
        {visible && renderMethod({ children, render }, width)}
      </EventListener>
    );
  }
}

export default Responsive;
