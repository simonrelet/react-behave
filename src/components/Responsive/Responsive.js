import debounce from 'lodash.debounce'
import isEqual from 'lodash.isequal'
import memoize from 'memoize-one'
import PropTypes from 'prop-types'
import React from 'react'
import getScreenSize from '../../core/getScreenSize'

function smallerOrEqualTo(screenSizes, refScreenSize, screenSize) {
  return screenSizes[screenSize] <= screenSizes[refScreenSize]
}

function biggerOrEqualTo(screenSizes, refScreenSize, screenSize) {
  return screenSizes[screenSize] >= screenSizes[refScreenSize]
}

/**
 * [props-minimum]: #minimum-string-optional
 * [props-maximum]: #maximum-string-optional
 * [props-screensizes]: #screensizes-object-optional
 *
 * Render content depending on the screen size.
 *
 * ## Usage
 *
 * ```jsx
 * import React from 'react'
 * import { Responsive } from 'react-behave'
 *
 * const screenSizes = {
 *   xs: 0,
 *   sm: 600,
 *   md: 960,
 *   lg: 1280,
 *   xl: 1920,
 * }
 *
 * class App extends React.Component {
 *   render() {
 *     return (
 *       <Responsive minimum="md" screenSizes={screenSizes}>
 *         <p>Only visible on medium and bigger screens</p>
 *       </Responsive>
 *     )
 *   }
 * }
 * ```
 */
class Responsive extends React.Component {
  getScreenSize = memoize(getScreenSize, isEqual)

  componentDidMount() {
    window.addEventListener('resize', this.handleResize)
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize)
    this.handleResize.cancel()
  }

  handleResize = debounce(() => this.forceUpdate(), this.props.resizeInterval)

  render() {
    const { children, maximum, minimum, screenSizes } = this.props
    const screenSize = this.getScreenSize(screenSizes, window.innerWidth)
    let visible = true

    if (minimum) {
      visible = biggerOrEqualTo(screenSizes, minimum, screenSize)
    }

    if (visible && maximum) {
      visible = smallerOrEqualTo(screenSizes, maximum, screenSize)
    }

    if (visible) {
      if (typeof children === 'function') {
        return children(screenSize, window.innerWidth)
      }

      return children
    }

    return null
  }
}

Responsive.propTypes = {
  /**
   * _Parameters_: `screenSize: String`, `width: Number`
   *
   * Invoked to render the children.
   *
   * The children will only be rendered if the current width is satisfied by [`props.maximum`][props-maximum] and [`props.minimum`][props-minimum].
   *
   * Example:
   *
   * ```jsx
   * import React from 'react'
   * import { Responsive } from 'react-behave'
   *
   * const screenSizes = {
   *   // [...]
   * }
   *
   * function LargeScreen() {
   *   return (
   *     <Responsive minimum="lg" screenSizes={screenSizes}>
   *       {screenSize => <p>The screen size is '{screenSize}'</p>}
   *     </Responsive>
   *   )
   * }
   * ```
   *
   * For convenience, `props.children` can also be a React element.
   * See [Usage](#usage).
   */
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]).isRequired,

  /**
   * Maximum screen width.
   * Must be one of the keys of [`props.screenSizes`][props-screensizes].
   */
  maximum: PropTypes.string,

  /**
   * Minimum screen width.
   * Must be one of the keys of [`props.screenSizes`][props-screensizes].
   */
  minimum: PropTypes.string,

  /**
   * The minimum interval between two resizes.
   */
  resizeInterval: PropTypes.number,

  /**
   * The screen sizes to use.
   * Each key is the name of a screen size that will be used by [`props.maximum`][props-maximum] and [`props.minimum`][props-minimum], and the value is the starting width in pixels of this range.
   *
   * For example:
   *
   * ```js
   * const screenSizes = {
   *   sm: 0,
   *   md: 960,
   *   lg: 1280,
   * }
   * ```
   *
   * Define the following screen sizes:
   *
   * - `'sm'`: [0, 960[
   * - `'md'`: [960, 1280[
   * - `'lg'`: [1280, âˆž[
   */
  screenSizes: PropTypes.object,
}

Responsive.defaultProps = {
  // Corresponds to 10 frames at 60 Hz
  resizeInterval: 166,
  screenSizes: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1280,
    xl: 1920,
  },
}

export default Responsive
