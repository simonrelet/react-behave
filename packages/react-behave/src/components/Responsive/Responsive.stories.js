import { object, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Responsive from './Responsive';

const defaultScreenSizes = Responsive.defaultProps.screenSizes;

const stories = storiesOf('Responsive', module);

stories.add('Render function', () => {
  const screenSizes = object('Screen sizes', defaultScreenSizes);
  const screenSizesOptions = Object.keys(screenSizes).reduce(
    (acc, size) => ({ ...acc, [size]: size }),
    { '': 'None' },
  );

  return (
    <Responsive
      maximum={select('Maximum', screenSizesOptions, '')}
      minimum={select('Minimum', screenSizesOptions, '')}
      screenSizes={screenSizes}
      render={width => (
        <div>
          <h3>The width is '{width}'.</h3>
          <em>Use the "Viewport" and "Knobs" tabs bellow to play around.</em>
        </div>
      )}
    />
  );
});

stories.add('Render children', () => (
  <Responsive maximum="md" minimum="sm">
    <p>I'm only visible on 'sm' and 'md' screens</p>
  </Responsive>
));
