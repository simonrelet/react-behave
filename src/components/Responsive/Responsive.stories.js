import { object, select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React, { Component } from 'react'
import { watchResize } from '../../core/watchResize'
import { getScreenSize } from '../../core/getScreenSize'
import { Responsive } from './Responsive'

const defaultScreenSizes = Responsive.defaultProps.screenSizes

const stories = storiesOf('Responsive', module)

stories.add('Render function', () => {
  const screenSizes = object('Screen sizes', defaultScreenSizes)
  const screenSizesOptions = Object.keys(screenSizes).reduce(
    (acc, size) => ({ ...acc, [size]: size }),
    { '': 'None' },
  )

  return (
    <Responsive
      maximum={select('Maximum', screenSizesOptions, '')}
      minimum={select('Minimum', screenSizesOptions, '')}
      screenSizes={screenSizes}
    >
      {screenSize => (
        <div>
          <h3>
            The screen size is '{screenSize}
            '.
          </h3>
          <em>Use the "Viewport" and "Knobs" tabs bellow to play around.</em>
        </div>
      )}
    </Responsive>
  )
})

stories.add('Render children', () => (
  <Responsive maximum="md" minimum="sm">
    <p>I'm only visible on 'sm' and 'md' screens</p>
  </Responsive>
))

class LocalResponsive extends Component {
  ref = React.createRef()
  stopWatching = null

  componentDidMount() {
    this.stopWatching = watchResize(this.ref.current, () => this.forceUpdate())
  }

  componentWillUnmount() {
    this.stopWatching()
  }

  getSize() {
    if (this.ref.current) {
      const { height, width } = this.ref.current.getBoundingClientRect()
      return {
        width: Math.floor(width),
        height: Math.floor(height),
      }
    }

    return {
      width: 0,
      height: 0,
    }
  }

  getScreenSize(width) {
    return getScreenSize(Responsive.defaultProps.screenSizes, width)
  }

  render() {
    const { width, height } = this.getSize()
    return (
      <div ref={this.ref}>
        <ul>
          <li>
            Width: {width}
            px
          </li>
          <li>
            Height: {height}
            px
          </li>
          <li>Screen size: {this.getScreenSize(width)}</li>
        </ul>
      </div>
    )
  }
}

stories.add('Local responsiveness', () => <LocalResponsive />)
