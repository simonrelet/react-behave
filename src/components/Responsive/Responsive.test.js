import React from 'react'
import { render } from 'react-testing-library'
import { Responsive } from './Responsive'

describe('<Responsive />', () => {
  let initialInnerWidth

  beforeEach(() => {
    initialInnerWidth = window.innerWidth
    jest.useFakeTimers()
  })

  afterEach(() => {
    window.innerWidth = initialInnerWidth
  })

  describe('<Responsive children />', () => {
    function createProps(props) {
      return {
        children: <p>Hello</p>,
        ...props,
      }
    }

    it('renders its content when there is no constraints', () => {
      const props = createProps()
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('renders its content when using `minimum`', () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.md
      const props = createProps({ minimum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `minimum`", () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.md - 1
      const props = createProps({ minimum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toBeNull()
    })

    it('renders its content when using `maximum`', () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.md
      const props = createProps({ maximum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `maximum`", () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.lg
      const props = createProps({ maximum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toBeNull()
    })

    it('renders its content when using `maximum` and `minimum`', () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.md
      const props = createProps({ minimum: 'sm', maximum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `maximum` and `minimum`", () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.xl
      const props = createProps({ minimum: 'sm', maximum: 'lg' })
      const { container } = render(<Responsive {...props} />)
      expect(container.firstChild).toBeNull()
    })

    it('listen to window resizes', () => {
      window.innerWidth = Responsive.defaultProps.screenSizes.lg
      const props = createProps({ minimum: 'md' })
      const { container } = render(<Responsive {...props} />)

      window.innerWidth = Responsive.defaultProps.screenSizes.sm
      window.dispatchEvent(new Event('resize'))
      jest.runAllTimers()
      expect(container.firstChild).toBeNull()
    })
  })

  describe('<Responsive render />', () => {
    function createProps(props) {
      return {
        children: jest.fn(() => <p>Hello</p>),
        ...props,
      }
    }

    it('renders its content when there is no constraints', () => {
      const size = Responsive.defaultProps.screenSizes.md
      window.innerWidth = size
      const props = createProps()
      const { container } = render(<Responsive {...props} />)
      expect(props.children).toHaveBeenCalledWith('md', size)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('renders its content when using `minimum`', () => {
      const size = Responsive.defaultProps.screenSizes.md
      window.innerWidth = size
      const props = createProps({ minimum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).toHaveBeenCalledWith('md', size)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `minimum`", () => {
      const size = Responsive.defaultProps.screenSizes.md - 1
      window.innerWidth = size
      const props = createProps({ minimum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).not.toHaveBeenCalled()
      expect(container.firstChild).toBeNull()
    })

    it('renders its content when using `maximum`', () => {
      const size = Responsive.defaultProps.screenSizes.md
      window.innerWidth = size
      const props = createProps({ maximum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).toHaveBeenCalledWith('md', size)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `maximum`", () => {
      const size = Responsive.defaultProps.screenSizes.lg
      window.innerWidth = size
      const props = createProps({ maximum: 'md' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).not.toHaveBeenCalled()
      expect(container.firstChild).toBeNull()
    })

    it('renders its content when using `maximum` and `minimum`', () => {
      const size = Responsive.defaultProps.screenSizes.md
      window.innerWidth = size
      const props = createProps({ minimum: 'sm', maximum: 'lg' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).toHaveBeenCalledWith('md', size)
      expect(container.firstChild).toMatchSnapshot()
    })

    it("doesn't render its content when using `maximum` and `minimum`", () => {
      const size = Responsive.defaultProps.screenSizes.xl
      window.innerWidth = size
      const props = createProps({ minimum: 'sm', maximum: 'lg' })
      const { container } = render(<Responsive {...props} />)
      expect(props.children).not.toHaveBeenCalled()
      expect(container.firstChild).toBeNull()
    })
  })
})
