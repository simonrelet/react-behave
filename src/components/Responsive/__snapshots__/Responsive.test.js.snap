// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Responsive /> <Responsive children /> renders its content when there is no constraints 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive children /> renders its content when using \`maximum\` 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive children /> renders its content when using \`maximum\` and \`minimum\` 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive children /> renders its content when using \`minimum\` 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive render /> renders its content when there is no constraints 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive render /> renders its content when using \`maximum\` 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive render /> renders its content when using \`maximum\` and \`minimum\` 1`] = `
<p>
  Hello
</p>
`;

exports[`<Responsive /> <Responsive render /> renders its content when using \`minimum\` 1`] = `
<p>
  Hello
</p>
`;
